File: 2015/20_InfiniteElvesandInfiniteHouses/fedelf.js
/* eslint-disable linebreak-style */
// ======================================================================
// Infinite Elves and Infinite Houses
//   Advent of Code 2015 Day 20 -- Eric Wastl -- https://adventofcode.com
//
// JavaScript implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                           f e d e l f . j s
//
// A solver for the Advent of Code 2015 Day 20 problem
// ======================================================================

// ----------------------------------------------------------------------
//                                                                 import
// ----------------------------------------------------------------------

// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------

// ======================================================================
//                                                                 Fedelf
// ======================================================================

class Fedelf {
  // Object for Infinite Elves and Infinite Houses

  constructor(options) {
    // Create a Fedelf object

    // 1. Set the initial values
    this.text = options.text === undefined ? null : options.text;
    this.part2 = options.part2 === undefined ? false : options.part2;
    this.presents = null;

    // 2. Process text (if any)
    if (this.text !== null) {
      this.presents = parseInt(this.text[0], 10);
    }
  }

  static inHouse(hnum) {
    // 1. Start with nothing
    let presents = 0;
    // 2. Loop for all possible elves
    for (let elf = 1; elf <= hnum; elf += 1) {
      // 3. If the elf delivers presents to this house, do so
      if (hnum % elf === 0) {
        presents += elf;
      }
    }
    // 4. Return the number of presents
    return presents * 10;
  }

  static firstHouse(presents) {
    // 1. Loop for all the possible houses
    for (let house = 1; house <= presents; house += 1) {
      // 2. Determine how many presents were delivered to that house
      const received = Fedelf.inHouse(house);
      // 3. Are there enough presents?
      if (received >= presents) {
        return house;
      }
    }
    // 4. So sorry
    return null;
  }

  partOne(options) {
    // Returns the solution for part one

    // 0. Function arguments
    // eslint-disable-next-line no-unused-vars
    const verbose = options.verbose === undefined ? false : options.verbose;
    // eslint-disable-next-line no-unused-vars
    const limit = options.limit === undefined ? 0 : options.limit;

    // 1. Return the solution for part one
    return Fedelf.firstHouse(this.presents);
  }

  static inHouse2(hnum, elves) {
    // 1. Start with nothing
    let presents = 0;
    const newElves = { ...elves };
    // 2. Loop for all possible elves
    for (let elf = 1; elf <= hnum; elf += 1) {
      // 3. If the elf could deliver presents to this house ...
      if (hnum % elf === 0) {
        // 4. Check if the elf is off the clock (already delivered 50 presents)
        if (elf in newElves) {
          newElves[elf] += 1;
        } else {
          newElves[elf] = 1;
        }
        if (newElves[elf] <= 50) {
          presents += elf;
        }
      }
    }
    // 5. Elf 'presents' has the number of presents
    newElves.presents = presents * 11;
    return newElves;
  }

  static firstHouse2(presents) {
    // 1. Reset the elves
    const elves = {};
    // 1. Loop for all the possible houses
    for (let house = 1; house <= presents; house += 1) {
      // 2. Add the elf that starts with this house
      elves[house] = 0;
      // 3. Determine how many presents were delivered to that house
      // 3a. Start with nothing
      let delivered = 0;
      // 3b. Loop for all possible elves
      Object.keys(elves).forEach((elf) => {
        // 3c. If the elf could deliver presents to this house ...
        if (house % elf === 0) {
          // 3d. Deliver the present
          delivered += parseInt(elf, 10);
          elves[elf] += 1;
          // 3e. If this the last one present for this elf, his/her work is done
          if (elves[elf] === 50) {
            delete elves[elf];
          }
        }
      });
      delivered *= 11;
      // eslint-disable-next-line no-console
      console.log(`house ${house} gets ${delivered}`);
      if (delivered >= presents) {
        return house;
      }
    }
    // 4. So sorry
    return null;
  }

  partTwo(options) {
    // Returns the solution for part two

    // 0. Function arguments
    // eslint-disable-next-line no-unused-vars
    const verbose = options.verbose === undefined ? false : options.verbose;
    // eslint-disable-next-line no-unused-vars
    const limit = options.limit === undefined ? 0 : options.limit;

    // 1. Return the solution for part two
    return Fedelf.firstHouse2(this.presents);
  }
}

module.exports.Fedelf = Fedelf;
// ======================================================================
// end                         f e d e l f . j s                      end
// ======================================================================

File: 2015/20_InfiniteElvesandInfiniteHouses/fedelf.py
# ======================================================================
# Infinite Elves and Infinite Houses
#   Advent of Code 2015 Day 20 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                         f e d e l f . p y
# ======================================================================
"A solver for the Advent of Code 2015 Day 20 puzzle"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
from collections import defaultdict

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------

# ======================================================================
#                                                                 Fedelf
# ======================================================================


class Fedelf(object):   # pylint: disable=R0902, R0205
    "Object for Infinite Elves and Infinite Houses"

    def __init__(self, text=None, part2=False):

        # 1. Set the initial values
        self.part2 = part2
        self.text = text
        self.goal = 0

        # 2. Process text (if any)
        if text is not None and len(text) > 0:
            self.goal = int(text[0])

    def fast_deliver(self, verbose=False):

        # 1. Get a whole bunch of houses
        houses = defaultdict(int)

        # 2. Loop for an infinate number of elves
        for elf in range(1, self.goal):

            # 3. Deliver to all of the houses
            for house in range(elf, self.goal, elf):

                # 4. Deliver presents to this house
                houses[house] += elf * 10

            # 5. Have we got enough presents delivered?
            if houses[elf] >= self.goal:
                return elf

            # 6, Verbose ouput
            if verbose and 0 == elf % 1000:
                print(elf, houses[elf])

        # 7. Didn't work
        return None

    def slow_deliver(self, verbose=False):
        "Deliver presents until we reach a house with the goal"

        # 1. Loop for a infinate number of houses
        house = 0
        presents = 0
        while presents < self.goal:

            # 2. Advance to the next house
            house += 1
            presents = 0

            # 3. Loop for all of the elves
            for elf in range(1, house + 1):

                # 4. Will this elf make a delivery?  If so accumulate presents
                if 0 == house % elf:
                    presents += elf * 10

            # 5, Verbose ouput
            if verbose and 0 == house % 1000:
                print(house, presents)

        # 6. Return the house number
        return house

    def deliver2(self, verbose=False):

        # 1. Get a whole bunch of houses
        houses = defaultdict(int)

        # 2. Loop for an infinate number of elves
        for elf in range(1, self.goal):

            # 3. Deliver to 50 houses
            for house in range(elf, min(1 + 50 * elf, self.goal), elf):

                # 4. Deliver presents to this house
                houses[house] += elf * 11

            # 5. Have we got enough presents delivered?
            if houses[elf] >= self.goal:
                return elf

            # 6, Verbose ouput
            if verbose and 0 == elf % 1000:
                print(elf, houses[elf])

        # 7. Didn't work
        return None

    def part_one(self, verbose=False, limit=0):
        "Returns the solution for part one"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part one
        return self.fast_deliver(verbose=verbose)

    def part_two(self, verbose=False, limit=0):
        "Returns the solution for part two"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part two
        return self.deliver2(verbose=verbose)


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                        f e d e l f . p y                       end
# ======================================================================
